File: 1339/main.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    MOD = 1e9 + 7

    def maxProduct(self, root: Optional[TreeNode]) -> int:
        self.sums = list[int]()

        treeSum = self.dfsSum(root)
        minDelta = treeSum

        for sum in self.sums:
            delta = abs(treeSum - 2 * sum)

            if minDelta > delta:
                minDelta = delta

        p1 = ((treeSum - minDelta) / 2) % Solution.MOD
        p2 = (treeSum - p1) % Solution.MOD

        return int((p1 * p2) % Solution.MOD)

    def dfsSum(self, node: Optional[TreeNode]) -> int:
        if not node:
            return 0

        sum = node.val
        sum += self.dfsSum(node.left)
        sum += self.dfsSum(node.right)
        self.sums.append(sum)

        return sum

File: 1339/main.ts
function maxProduct(root: TreeNode | null): number {
	const MOD = 1e9 + 7;
	const sums = new Array<number>();

	const dfsSum = (node: TreeNode | null): number => {
		if (node === null) return 0;

		let sum = node.val;
		sum += dfsSum(node.left);
		sum += dfsSum(node.right);

		sums.push(sum);

		return sum;
	};

	const treeSum = dfsSum(root);
	let minDelta = treeSum;

	for (const sum of sums) {
		const delta = Math.abs(treeSum - 2 * sum);

		if (minDelta > delta) {
			minDelta = delta;
		}
	}

	const p1 = ((treeSum - minDelta) / 2) % MOD;
	const p2 = (treeSum - p1) % MOD;

	return (p1 * p2) % MOD;
}

/*   *   *   *   *   *   *   *   *   *   */
/*   *   *   *   *   *   *   *   *   *   */

class TreeNode {
	val: number;
	left: TreeNode | null;
	right: TreeNode | null;
	constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
		this.val = val === undefined ? 0 : val;
		this.left = left === undefined ? null : left;
		this.right = right === undefined ? null : right;
	}
}
